[HtmlSanitizer] Introduce HtmlSanitizer component

tgalopin opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix https://github.com/symfony/symfony/issues/44144
| License       | MIT
| Doc PR        | -

Following discussion on https://github.com/symfony/symfony/issues/44144, this PR proposes to add a new HtmlSanitizer component to Symfony.

I refactored the infrastructure of my original [html-sanitizer](https://github.com/tgalopin/html-sanitizer) to better follow the [W3C Standard Proposal](https://wicg.github.io/sanitizer-api/) about sanitizers.

The sanitizer can be used as follow:

```php
use Symfony\Component\HtmlSanitizer\HtmlSanitizerConfig;
use Symfony\Component\HtmlSanitizer\HtmlSanitizer;

// By default, an element not added to the allowed or blocked elements
// will be dropped, including its children
$config = (new HtmlSanitizerConfig())
    // Allow "safe" elements and attributes. All scripts will be removed
    // as well as other dangerous behaviors like CSS injection
    ->allowSafeElements()

    // Allow all static elements and attributes from the W3C Sanitizer API
    // standard. All scripts will be removed but the output may still contain
    // other dangerous behaviors like CSS injection (click-jacking), CSS
    // expressions, ...
    ->allowAllStaticElements()

    // Allow the "div" element and no attribute can be on it
    ->allowElement('div')

    // Allow the "a" element, and the "title" attribute to be on it
    ->allowElement('a', ['title'])

    // Allow the "span" element, and any attribute from the Sanitizer API is allowed
    // (see https://wicg.github.io/sanitizer-api/#default-configuration)
    ->allowElement('span', '*')

    // Block the "section" element: this element will be removed but
    // its children will be retained
    ->blockElement('section')

    // Drop the "div" element: this element will be removed, including its children
    ->dropElement('div')

    // Allow the attribute "title" on the "div" element
    ->allowAttribute('title', ['div'])

    // Allow the attribute "data-custom-attr" on all currently allowed elements
    ->allowAttribute('data-custom-attr', '*')

    // Drop the "data-custom-attr" attribute from the "div" element:
    // this attribute will be removed
    ->dropAttribute('data-custom-attr', ['div'])

    // Drop the "data-custom-attr" attribute from all elements:
    // this attribute will be removed
    ->dropAttribute('data-custom-attr', '*')

    // Forcefully set the value of all "rel" attributes on "a"
    // elements to "noopener noreferrer"
    ->forceAttribute('a', 'rel', 'noopener noreferrer')

    // Transform all HTTP schemes to HTTPS
    ->forceHttpsUrls()

    // Configure which schemes are allowed in links (others will be dropped)
    ->allowedLinkSchemes(['https', 'http', 'mailto'])

    // Configure which hosts are allowed in links (by default all are allowed)
    ->allowedLinkHosts(['symfony.com', 'example.com'])

    // Allow relative URL in links (by default they are dropped)
    ->allowRelativeLinks()

    // Configure which schemes are allowed in img/audio/video/iframe (others will be dropped)
    ->allowedMediaSchemes(['https', 'http'])

    // Configure which hosts are allowed in img/audio/video/iframe (by default all are allowed)
    ->allowedMediaHosts(['symfony.com', 'example.com'])

    // Allow relative URL in img/audio/video/iframe (by default they are dropped)
    ->allowRelativeMedias()

    // Configure a custom attribute sanitizer to apply custom sanitization logic
    // ($attributeSanitizer instance of AttributeSanitizerInterface)
    ->withAttributeSanitizer($attributeSanitizer)

    // Unregister a previously registered attribute sanitizer
    // ($attributeSanitizer instance of AttributeSanitizerInterface)
    ->withoutAttributeSanitizer($attributeSanitizer)
;

$sanitizer = new HtmlSanitizer($config);

// Sanitize a given string, using the configuration provided, and in the
// "body" context (tags only allowed in <head> will be removed)
$sanitizer->sanitize($userInput);

// Sanitize the given string for a usage in a <head> tag
$sanitizer->sanitizeFor('head', $userInput);

// Sanitize the given string for a usage in another tag
$sanitizer->sanitizeFor('title', $userInput); // Will encode as HTML entities
$sanitizer->sanitizeFor('textarea', $userInput); // Will encode as HTML entities
$sanitizer->sanitizeFor('div', $userInput); // Will sanitize as body
$sanitizer->sanitizeFor('section', $userInput); // Will sanitize as body
// ...
```

The component is working and well tested.

I did some very basic performance testing and it seems to be **~ 2 times faster** than my original library, mostly by relying on much less objects and more on native primitives. My original library was already as fast as the fastest PHP HTML sanitizers, meaning this component may well be the fastest sanitizer in the PHP ecosystem (to better check of course!). That's exciting :) !

I still have on my TODO list:

- [x] Review and update the annotations in all classes and methods (especially to mark the component as experimental) ;
- [ ] Add the FameworkBundle integration to use it in Symfony apps

Feel free to review the component while I work on finalizing it and the FameworkBundle integration!

tgalopin committed:
[HtmlSanitizer] Introduce HtmlSanitizer component

derrabus commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizer.php:
```diff
@@ -0,0 +1,132 @@
+    {
+        $this->config = clone $config;
+        $this->maxInputLength = $maxInputLength;
+        $this->parser = $parser ?: new MastermindsParser();
```
```suggestion
        $this->parser = $parser ?? new MastermindsParser();
```

xorik commented:
Hi.

First of all, thank you for you effort, this is really cool idea!

Second: I'm not a Symony team member, so feel free to ignore me.

My concern is that configuring will create a bunch of temporary objects. Maybe it's better to make a separate builder class, something like:

```php
<?php

$config = (new HtmlSanitizerConfigBuilder())
    ->allowElement('div')
    ->allowElement('a', ['title'])
    ->getConfig()
;
$sanitizer = new HtmlSanitizer($config);
```
And call constructor only in the `getConfig` method, and not on every step, so it will be less CPU intensive. But maybe it's not required, if this will be somehow optimized in DI to a single constructor call.

derrabus commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizer.php:
```diff
@@ -0,0 +1,132 @@
+        // Other context: build a DOM visitor
+        if (!isset($this->domVisitors[$context])) {
+            $this->domVisitors[$context] = $this->createDomVisitorForContext($context);
+        }
```
```suggestion
        $this->domVisitors[$context] ??= $this->createDomVisitorForContext($context);
```

derrabus commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizer.php:
```diff
@@ -0,0 +1,132 @@
+
+        try {
+            $parsed = $this->parser->parse($input);
+        } catch (\Exception) {
```
Can we be more specific about what kind of exception we expect here?

derrabus commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,349 @@
+        $self = clone $this;
+
+        // Unblock the element is necessary
+        if (isset($self->blockedElements[$element])) {
+            unset($self->blockedElements[$element]);
+        }
```
```suggestion
        unset($self->blockedElements[$element]);
```

`unset()` should be idempotent. No need to safe-guard the call.

derrabus commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,349 @@
+ */
+class HtmlSanitizerConfig
+{
+    // Elements that should be removed but their children should be retained
```
Please use PHPDoc syntax to document properties.

derrabus commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,349 @@
+class HtmlSanitizerConfig
+{
+    // Elements that should be removed but their children should be retained
+    private array $blockedElements = [];
```
Can we be more specific on the kind of elements that the array properties in this class may contain?

derrabus commented on src/Symfony/Component/HtmlSanitizer/Parser/Exception/ParsingFailedException.php:
```diff
@@ -0,0 +1,34 @@
+
+    public function __construct(ParserInterface $parser, \Throwable $previous = null)
+    {
+        parent::__construct('HTML parsing failed using parser '.\get_class($parser), 0, $previous);
```
```suggestion
        parent::__construct('HTML parsing failed using parser '.\get_debug_type($parser), 0, $previous);
```

derrabus commented on src/Symfony/Component/HtmlSanitizer/composer.json:
```diff
@@ -0,0 +1,33 @@
+        "league/uri": "^6.5",
+        "masterminds/html5": "^2.4",
+        "psr/log": "^1.0|^2.0|^3.0",
+        "symfony/string": "^5.0|^6.0"
```
```suggestion
        "symfony/string": "^5.4|^6.0"
```

We've bumped all Symfony dependencies to `^5.4|^6.0` on the 6.0  branch. We should do the same for new components.

derrabus commented on src/Symfony/Component/HtmlSanitizer/Visitor/Node/TextNode.php:
```diff
@@ -0,0 +1,43 @@
+ */
+class TextNode implements NodeInterface
+{
+    private NodeInterface $parentNode;
+    private string $text;
+
+    public function __construct(NodeInterface $parentNode, string $text)
+    {
+        $this->parentNode = $parentNode;
+        $this->text = $text;
+    }
```
```suggestion
    public function __construct(
        private NodeInterface $parentNode,
        private string $text
    ) {
    }
```

nicolas-grekas replied:
with trailing coma :)

derrabus commented on src/Symfony/Component/HtmlSanitizer/TextSanitizer/UrlSanitizer.php:
```diff
@@ -0,0 +1,115 @@
+            return null;
+        }
+
+        try {
+            $parsed = parse($url);
+        } catch (Exception) {
+            return null;
+        }
+
+        return $parsed;
```
```suggestion
        try {
            return parse($url);
        } catch (Exception) {
            return null;
        }
```

GromNaN commented on src/Symfony/Component/HtmlSanitizer/Parser/MastermindsParser.php:
```diff
@@ -0,0 +1,32 @@
+ *
+ * @final
+ */
+class MastermindsParser implements ParserInterface
```
Any reason why this class is not actually `final`?

stof commented on src/Symfony/Component/HtmlSanitizer/Visitor/DomVisitor.php:
```diff
@@ -0,0 +1,134 @@
+
+    private function visitNode(\DOMNode $domNode, Cursor $cursor)
+    {
+        $nodeName = u($domNode->nodeName)->lower()->toString();
```
Do we actually need to use symfony/string to lowercase the node name ? Node names are ASCII only AFAIK.

OskarStark replied:
If this component is only used once, I would drop this dep 👍🏻

GromNaN replied:
@OskarStark used in several other places (see `HtmlSanitizer`).

tgalopin replied:
My reasoning was the ability of the String component to handle encoding better than native functions. The component is used in 3 places:

* Right here
* https://github.com/symfony/symfony/pull/44681/files#diff-5f51ebdb3ffcee374edaa31e4bc927becaefb2122ff4ad3069d4ba333ac32333R65 to limit the max length of the string for DoS prevention
* https://github.com/symfony/symfony/pull/44681/files#diff-3bb0fd9c4da8e8a650092f0ec099105f1201c3da2fad4ea525520495ee149224R136 to lower attributes names

We may be fine by removing the component, I just wanted to ensure a good handling of various encodings. I'm fine with both scenarios.

tgalopin replied:
Dependency removed

stof replied:
attribute names are also ASCII only AFAIK.
And for DoS prevention, a limit based on the byte length rather than the character length is OK.

stof commented on src/Symfony/Component/HtmlSanitizer/Visitor/AttributeSanitizer/AttributeSanitizerInterface.php:
```diff
@@ -0,0 +1,23 @@
+
+interface AttributeSanitizerInterface
+{
+    public function getSupportedElements(): array;
```
Please add documentation for those interfaces:

- method description explaining what the method is about
- phpdoc describing the type of arrays

stof commented on src/Symfony/Component/HtmlSanitizer/TextSanitizer/UrlSanitizer.php:
```diff
@@ -0,0 +1,115 @@
+
+use function League\Uri\build;
+use League\Uri\Exception;
+use function League\Uri\parse;
```
If we start adding use statements for functions, I suggest sorting the use statements by type

OskarStark replied:
Should be a job for php-ca-fixer

stof replied:
php-cs-fixer supports that, but the Symfony config does not currently enable the use statement sorting IIRC

nicolas-grekas replied:
I'm for removing all `use function` and doing `use League\Uri` + `Uri\parse()` instead.

stof commented on src/Symfony/Component/HtmlSanitizer/README.md:
```diff
@@ -0,0 +1,91 @@
+Example usage:
```
The readme should contain the info present in all component readme files.

stof commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,349 @@
+    // Should the URL in the sanitized document be transformed to HTTPS if they are using HTTP
+    private bool $forceHttpsUrls = false;
+
+    // Sanitizers that should be applied to specific attributes in addition to standard sanitation
```
Typo here (`sanitation` is not a valid word).

tgalopin replied:
I think sanitation is actually the right word: https://www.linguee.fr/anglais-francais/traduction/sanitation.html

tgalopin replied:
Just changed to sanitization, both actually exist but sanitization makes more sense :) .

stof commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,349 @@
+    // Should the URL in the sanitized document be transformed to HTTPS if they are using HTTP
+    private bool $forceHttpsUrls = false;
+
+    // Sanitizers that should be applied to specific attributes in addition to standard sanitation
```
The property description should be in the doc comment of the property rather than in a separate comment

stof commented on src/Symfony/Component/HtmlSanitizer/Parser/Exception/ParsingFailedException.php:
```diff
@@ -0,0 +1,34 @@
+        $this->parser = $parser;
+    }
+
+    public function getParser(): ParserInterface
```
Why exposing the parser in the exception ?

OskarStark commented on src/Symfony/Component/HtmlSanitizer/CHANGELOG.md:
```diff
@@ -0,0 +1,7 @@
+6.1.0
+-----
+
+ * added the component as experimental
```
```suggestion
 * Add the component as experimental
```

GromNaN replied:
Yes, this is how changelogs are written now.
https://github.com/symfony/symfony/blob/5.4.0/src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md#54

OskarStark commented on src/Symfony/Component/HtmlSanitizer/Parser/MastermindsParser.php:
```diff
@@ -0,0 +1,32 @@
+ */
+class MastermindsParser implements ParserInterface
+{
+    public function parse(string $html): \DOMNode
```
Inheritdoc? Or is this not needed?

GromNaN replied:
Only necessary when there is an additional comment.

OskarStark commented on src/Symfony/Component/HtmlSanitizer/Tests/HtmlSanitizerCustomTest.php:
```diff
@@ -0,0 +1,404 @@
+use Symfony\Component\HtmlSanitizer\HtmlSanitizerConfig;
+use Symfony\Component\HtmlSanitizer\Visitor\AttributeSanitizer\AttributeSanitizerInterface;
+
+class HtmlSanitizerCustomTest extends TestCase
```
Final please

nicolas-grekas replied:
Nope, we don't use `final` on test cases, for the same reasons we don't use `void` either.

OskarStark replied:
Ok, cannot check right now, but as far as I remember, we do use it for notifier tests

OskarStark commented on src/Symfony/Component/HtmlSanitizer/Tests/HtmlSanitizerConfigTest.php:
```diff
@@ -0,0 +1,279 @@
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\HtmlSanitizer\HtmlSanitizerConfig;
+
+class HtmlSanitizerConfigTest extends TestCase
```
Final please

OskarStark commented on src/Symfony/Component/HtmlSanitizer/Tests/HtmlSanitizerStandardTest.php:
```diff
@@ -0,0 +1,432 @@
+use Symfony\Component\HtmlSanitizer\HtmlSanitizer;
+use Symfony\Component\HtmlSanitizer\HtmlSanitizerConfig;
+
+class HtmlSanitizerStandardTest extends TestCase
```
Final please

OskarStark commented on src/Symfony/Component/HtmlSanitizer/Tests/Parser/MastermindsParserTest.php:
```diff
@@ -0,0 +1,27 @@
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\HtmlSanitizer\Parser\MastermindsParser;
+
+class MastermindsParserTest extends TestCase
```
Final please

OskarStark commented on src/Symfony/Component/HtmlSanitizer/Tests/TextSanitizer/StringSanitizerTest.php:
```diff
@@ -0,0 +1,48 @@
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\HtmlSanitizer\TextSanitizer\StringSanitizer;
+
+class StringSanitizerTest extends TestCase
```
Final please

OskarStark commented on src/Symfony/Component/HtmlSanitizer/Tests/TextSanitizer/UrlSanitizerTest.php:
```diff
@@ -0,0 +1,783 @@
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\HtmlSanitizer\TextSanitizer\UrlSanitizer;
+
+class UrlSanitizerTest extends TestCase
```
Final please

OskarStark commented on src/Symfony/Component/HtmlSanitizer/Tests/TextSanitizer/UrlSanitizerTest.php:
```diff
@@ -0,0 +1,783 @@
+     */
+    public function testSanitize($input, $allowedSchemes, $allowedHosts, $forceHttps, $allowRelative, $expected)
+    {
+        $this->assertEquals($expected, UrlSanitizer::sanitize($input, $allowedSchemes, $forceHttps, $allowedHosts, $allowRelative));
```
AssertSame possible?

OskarStark commented on src/Symfony/Component/HtmlSanitizer/Tests/TextSanitizer/UrlSanitizerTest.php:
```diff
@@ -0,0 +1,783 @@
+    /**
+     * @dataProvider provideSanitize
+     */
+    public function testSanitize($input, $allowedSchemes, $allowedHosts, $forceHttps, $allowRelative, $expected)
```
What about argument types?

OskarStark commented on src/Symfony/Component/HtmlSanitizer/Visitor/AttributeSanitizer/AHrefSanitizer.php:
```diff
@@ -0,0 +1,39 @@
+use Symfony\Component\HtmlSanitizer\HtmlSanitizerConfig;
+use Symfony\Component\HtmlSanitizer\TextSanitizer\UrlSanitizer;
+
+class AHrefSanitizer implements AttributeSanitizerInterface
```
Final?

Why is this sanitizer not internal, while the others are?

OskarStark commented on src/Symfony/Component/HtmlSanitizer/Visitor/DomVisitor.php:
```diff
@@ -0,0 +1,134 @@
+    private array $allowedNodes;
+    private HtmlSanitizerConfig $config;
+
+    /** @var AttributeSanitizer\AttributeSanitizerInterface[][][] */
```
Why the aliased namespace?

GromNaN replied:
Not an alias, but a sub-namespace of the current.

OskarStark replied:
Indeed, that's what I meant

OskarStark commented on src/Symfony/Component/HtmlSanitizer/Visitor/Model/Cursor.php:
```diff
@@ -0,0 +1,24 @@
+ *
+ * @final
+ */
+class Cursor
```
Why no real final?

OskarStark commented on src/Symfony/Component/HtmlSanitizer/Visitor/DomVisitor.php:
```diff
@@ -0,0 +1,134 @@
+use function Symfony\Component\String\u;
+
+/**
+ * The DomVisitor iterate over the parsed DOM tree and visit nodes using the NodeVisitor.
```
```suggestion
 * The DomVisitor iterates over the parsed DOM tree and visits nodes using the NodeVisitor.
```
?

OskarStark commented on src/Symfony/Component/HtmlSanitizer/Visitor/Node/NodeInterface.php:
```diff
@@ -0,0 +1,24 @@
+ */
+interface NodeInterface
+{
+    public function addChild(self $node): void;
```
Please describe the methods

OskarStark commented on src/Symfony/Component/HtmlSanitizer/phpunit.xml.dist:
```diff
@@ -0,0 +1,30 @@
+    </php>
+
+    <testsuites>
+        <testsuite name="Symfony String Component Test Suite">
```
Wrong name

tgalopin commented:
@xorik using a builder for the config could be an alternative indeed, but I think there is an advantage of using an immutable object and withers: the ability to store a "standard" configuration as a service and use it as a prototype to configure in your application. Using a builder would make this more complex while introducing a state in the service, which is IMO not a good idea.

Also, having a few more instantiated objects won't have any measurable performance impact, so I think we should stick to an immutable configuration object (immutability was proposed by @stof and @nicolas-grekas and I think that's a great idea).

@OskarStark I'm reviewing all your feedback (thanks for it!), I don't think we put final keywords on tests classes anywhere or did we? I feel like this is unnecessary as tests are excluded from the BC promise anyway.

Applying the other changes now :)

nicolas-grekas commented on src/Symfony/Component/HtmlSanitizer/Visitor/DomVisitor.php:
```diff
@@ -0,0 +1,150 @@
+
+        /** @var \DOMAttr $attribute */
+        foreach ($domNode->attributes as $attribute) {
+            $name = u($attribute->name)->lower()->toString();
```
from the rfc, lowercasing should happen only on A-Z, not on other unicode chars:
https://w3c.github.io/html-reference/terminology.html#case-insensitive

nicolas-grekas commented on src/Symfony/Component/HtmlSanitizer/CHANGELOG.md:
```diff
@@ -0,0 +1,7 @@
+CHANGELOG
+=========
+
+6.1.0
```
`6.1` only

nicolas-grekas commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizer.php:
```diff
@@ -0,0 +1,128 @@
+
+    public function __construct(HtmlSanitizerConfig $config, int $maxInputLength = 20000, ParserInterface $parser = null)
+    {
+        $this->config = clone $config;
```
no need to clone since $config is immutable

nicolas-grekas commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizer.php:
```diff
@@ -0,0 +1,128 @@
+
+    public function sanitizeFor(string $element, string $input): string
+    {
+        return $this->sanitizeWithContext(W3CReference::CONTEXTS_MAP[$element] ?? W3CReference::CONTEXT_BODY, $input);
```
not sure of the fallback to BODY since there is already a method for that. shouldn't we throw instead?

stof replied:
Well, most elements don't have special rules for that and need the same rule than body. this logic here avoids making `CONTEXTS_MAP` huge.

nicolas-grekas replied:
Makes sense thanks

nicolas-grekas commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizer.php:
```diff
@@ -0,0 +1,128 @@
+    private function isValidUtf8(string $html): bool
+    {
+        // preg_match() fails silently on strings containing invalid UTF-8.
+        return '' === $html || 1 === preg_match('/^./us', $html);
```
`'' === $html || preg_match('//u', $html);`

nicolas-grekas commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,373 @@
+     *
+     * Calling this method will override all previous configuration for allowed links schemes.
+     */
+    public function withAllowedLinksSchemes(array $allowedLinksSchemes): self
```
`: static` everywhere

nicolas-grekas commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,373 @@
+     */
+    public function withAllowedLinksSchemes(array $allowedLinksSchemes): self
+    {
+        $self = clone $this;
```
`$clone = `

nicolas-grekas commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,373 @@
+     *
+     * Calling this method will override all previous configuration for allowed links schemes.
+     */
+    public function withAllowedLinksHosts(?array $allowedLinksHosts): self
```
let's try removing the with prefix as much as possible since not all methods use it already?
eg `allowLinksHosts()` here?
no `s` to `Link` btw? (same to other similar methods)

nicolas-grekas commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerInterface.php:
```diff
@@ -0,0 +1,56 @@
+     * sanitizer will adapt its rules to only allow elements that are valid inside
+     * the given parent element.
+     *
+     * @param string $element A parent element name ("body", "head", "textarea", "div", ...).
```
no final dots

nicolas-grekas commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerInterface.php:
```diff
@@ -0,0 +1,56 @@
+     * the given parent element.
+     *
+     * @param string $element A parent element name ("body", "head", "textarea", "div", ...).
+     * @param string $input   the untrusted HTML input to sanitize
```
titlecase "The" (please check other phpdoc for similar fixes)

nicolas-grekas commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerInterface.php:
```diff
@@ -0,0 +1,56 @@
+     * @param string $element A parent element name ("body", "head", "textarea", "div", ...).
+     * @param string $input   the untrusted HTML input to sanitize
+     *
+     * @return string a safe HTML string that can be injected into document’s DOM
```
to be removed, the description of the method says it already

nicolas-grekas commented on src/Symfony/Component/HtmlSanitizer/Reference/W3CReference.php:
```diff
@@ -0,0 +1,383 @@
+ *
+ * @internal
+ */
+final class W3CReference
```
how did you generate this class? how will we update it if the w3c updates it?

tgalopin replied:
I added a test case to ensure it'll stay in sync with the standard.

nicolas-grekas commented on src/Symfony/Component/HtmlSanitizer/Visitor/AttributeSanitizer/AHrefSanitizer.php:
```diff
@@ -0,0 +1,42 @@
+use Symfony\Component\HtmlSanitizer\TextSanitizer\UrlSanitizer;
+
+/**
+ * @internal
```
why internal? no need to for classes that implement a public interface

nicolas-grekas commented on src/Symfony/Component/HtmlSanitizer/Visitor/AttributeSanitizer/MediaSrcSanitizer.php:
```diff
@@ -0,0 +1,42 @@
+use Symfony\Component\HtmlSanitizer\TextSanitizer\UrlSanitizer;
+
+/**
+ * @internal
```
might not be needed also

nicolas-grekas commented on src/Symfony/Component/HtmlSanitizer/Visitor/Node/Node.php:
```diff
@@ -0,0 +1,106 @@
+/**
+ * @author Titouan Galopin <galopintitouan@gmail.com>
+ *
+ * @internal
```
needed?

nicolas-grekas commented on src/Symfony/Component/HtmlSanitizer/Visitor/Node/DocumentNode.php:
```diff
@@ -0,0 +1,42 @@
+/**
+ * @author Titouan Galopin <galopintitouan@gmail.com>
+ *
+ * @internal
```
needed?

nicolas-grekas commented on src/Symfony/Component/HtmlSanitizer/Visitor/Node/BlockedNode.php:
```diff
@@ -0,0 +1,48 @@
+/**
+ * @author Titouan Galopin <galopintitouan@gmail.com>
+ *
+ * @internal
```
needed?

nicolas-grekas reviewed (commented):
some random notes :)

tgalopin commented:
I updated the PR for requested changes.

stof commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,373 @@
+     *
+     * @var string[]|null
+     */
+    private ?array $allowLinkHosts = null;
```
shouldn't those array properties be named `allowed*` rather than `allow*` ? they contain the list of allowed values

tgalopin replied:
Well, they describe a future sanitization, ie. this config "allows" certain hosts for links. Allowed would be a description of the content, allow is description of the aim. I'm not sure either are better TBH.

stof replied:
Well, I would keep `allow*()` as name for the configuration method. However, for the stored config and the getter, I would use `allowed` (like for `allowedElements` btw)

stof commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,373 @@
+    /**
+     * Sanitizers that should be applied to specific attributes in addition to standard sanitization.
+     *
+     * @var Visitor\AttributeSanitizer\AttributeSanitizerInterface[]
```
should use `AttributeSanitizerInterface[]` as there is a use statement

stof commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,373 @@
+
+        $clone->allowedElements[$element] = [];
+
+        $attrs = ('*' === $allowedAttributes) ? W3CReference::ATTRIBUTES : (array) $allowedAttributes;
```
should we allow using `*` inside the array too, in case we want to allow all standard attributes plus some custom ones ?

stof commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,373 @@
+        $droppedElements = ('*' === $droppedElements) ? array_keys($clone->allowedElements) : (array) $droppedElements;
+
+        foreach ($droppedElements as $element) {
+            if (isset($clone->allowedElements[$element][$attribute])) {
```
you can unused without this condition

stof commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,373 @@
+     * A list of element for which to drop this attribute can be passed as a second argument.
+     * Passing "*" will drop the attribute on all currently configured elements.
+     *
+     * Calling this method will override all previous configuration for the given attribute.
```
We need some documentation explaining that this method should be called after calls to `allowElement()` (otherwise the newly allowed element won't have the attribute dropped)

stof commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,373 @@
+        return $clone;
+    }
+
+    public function getAllowedElements(): array
```
the phpdoc should be added on all those methods to document the structure of arrays.

nicolas-grekas commented:
It'd be nice to be able to force an attribute on an element, eg rel=nofollow on `<a>`

stof commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,373 @@
+    }
+
+    /**
+     * @return Visitor\AttributeSanitizer\AttributeSanitizerInterface[]
```
`AttributeSanitizerInterface[]`

stof commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizer.php:
```diff
@@ -0,0 +1,128 @@
+
+    public function sanitizeFor(string $element, string $input): string
+    {
+        return $this->sanitizeWithContext(W3CReference::CONTEXTS_MAP[$element] ?? W3CReference::CONTEXT_BODY, $input);
```
given that element names are case insensitive in HTML, should we lowercase `$element` here ?

stof commented on src/Symfony/Component/HtmlSanitizer/Tests/Parser/MastermindsParserTest.php:
```diff
@@ -0,0 +1,27 @@
+
+class MastermindsParserTest extends TestCase
+{
+    public function testParseValid()
```
should we also have a test parsing an invalid HTML to assert the behavior ?

tgalopin replied:
The parser never throws exception (which is why I removed the exception mechanism I had originally). It will always return a string, and if there have been parsing errors it will provide them in an independent method call. If one wants to get these errors themselves, they could use a customer ParserInterface.

Testing for invalid HTML is therefore already done in the HtmlSanitizer tests, where invalid HTML tags are provided to ensure no unsafe output is generated.

stof commented on src/Symfony/Component/HtmlSanitizer/Tests/Reference/W3CReferenceTest.php:
```diff
@@ -0,0 +1,47 @@
+{
+    private const STANDARD_RESOURCES = [
+        'elements' => 'https://raw.githubusercontent.com/WICG/sanitizer-api/main/resources/baseline-element-allow-list.json',
+        'attributes' => 'https://raw.githubusercontent.com/WICG/sanitizer-api/main/resources/baseline-attribute-allow-list.json',
```
Comparing to the main branch of the repo in our testsuite will make our CI break in all branches when the W3C makes a change (and would force us to apply such updates in patch versions). Is it what we want ?

tgalopin replied:
That's what @nicolas-grekas proposed and the idea seems good to me. We may want to reconsider, it's up to debate.

stof commented on src/Symfony/Component/HtmlSanitizer/Tests/TextSanitizer/UrlSanitizerTest.php:
```diff
@@ -0,0 +1,783 @@
+            'http://example.com/你好你好' => ['scheme' => 'http', 'host' => 'example.com'],
+            'http://example.com/‥/foo' => ['scheme' => 'http', 'host' => 'example.com'],
+            'http://example.com/﻿/foo' => ['scheme' => 'http', 'host' => 'example.com'],
+            'http://example.com/‮/foo/‭/bar' => ['scheme' => 'http', 'host' => 'example.com'],
```
Github warns about bidirectional chars here. Can you write them using PHP escape sequences instead, which makes it clear that this is what the test case is about, and avoids such warning (marking the code as suspicious as a non-intended usage could be an attack) ?

stof commented on src/Symfony/Component/HtmlSanitizer/TextSanitizer/StringSanitizer.php:
```diff
@@ -0,0 +1,65 @@
+        // Some DB engines will transform UTF8 full-width characters their classical version
+        // if the data is saved in a non-UTF8 field
+        '＜' => '&#xFF1C;',
+        '＞' => '&#xFF1E;',
```
Those don't seem to be covered in tests

stof commented on src/Symfony/Component/HtmlSanitizer/TextSanitizer/UrlSanitizer.php:
```diff
@@ -0,0 +1,112 @@
+        return UriString::build($url);
+    }
+
+    public static function parse(string $url): ?array
```
why is this a public method ?

tgalopin replied:
Initially it was mostly to be able to test it independently from the sanitize method, I can close it and remove the tests if we don't think they are necessary (or at least put it as internal).

nicolas-grekas replied:
yep, better close this, it should not be part of the public API

nicolas-grekas replied:
oh, the class is internal already, so that's fine :)

stof commented on src/Symfony/Component/HtmlSanitizer/Visitor/DomVisitor.php:
```diff
@@ -0,0 +1,150 @@
+     * Registry of attributes sanitizers indexed by element name and attribute name for
+     * faster sanitization.
+     *
+     * @var AttributeSanitizerInterface[][][]
```
proper type is `array<string, array<string, AttributeSanitizerInterface[]>>`

stof commented on src/Symfony/Component/HtmlSanitizer/Visitor/DomVisitor.php:
```diff
@@ -0,0 +1,150 @@
+     */
+    private array $attributesSanitizers = [];
+
+    public function __construct(HtmlSanitizerConfig $config, array $elementsConfig)
```
The type of the `$elementsConfig` array should be documented (based on the phpdoc, it looks like `array<string, string[]|false>`)

stof commented on src/Symfony/Component/HtmlSanitizer/Visitor/DomVisitor.php:
```diff
@@ -0,0 +1,150 @@
+        // Otherwise, visit recursively
+        $this->enterNode($nodeName, $domNode, $cursor);
+        $this->visitChildren($domNode, $cursor);
+        $cursor->node = $cursor->node->getParent();
```
this assigns the returned value of type `NodeInterface|null` into a property typed as `NodeInterface` (btw, I find it weird that psalm does not complain about it)

stof commented on src/Symfony/Component/HtmlSanitizer/Visitor/Node/TextNode.php:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    public function addChild(NodeInterface $node): void
+    {
```
should this throw a LogicException ?

stof commented on src/Symfony/Component/HtmlSanitizer/composer.json:
```diff
@@ -0,0 +1,32 @@
+        "ext-dom": "*",
+        "league/uri": "^6.5",
+        "masterminds/html5": "^2.4",
+        "psr/log": "^1.0|^2.0|^3.0"
```
is the component actually using psr/log ?

tgalopin replied:
It was, it indeed doesn't anymore, let's remove it :)

stof commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,373 @@
+    {
+        $self = new self();
+
+        foreach (array_merge(W3CReference::BODY_ELEMENTS, W3CReference::HEAD_ELEMENTS) as $element) {
```
this is broken (see psalm detecting it too). Those constants are storing the element name as array key, not as array value.

ro0NL commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerInterface.php:
```diff
@@ -0,0 +1,47 @@
+ *
+ * @experimental
+ */
+interface HtmlSanitizerInterface
```
needed already?

tgalopin replied:
What do you mean? It'll be necessary when injecting the service won't it?

ro0NL replied:
i could inject `HtmlSanitizer`, i feel like a custom parser should be the main extension/testing entrypoint at first.

nicolas-grekas replied:
Why would this be a better design, from a SOLID pov?

ro0NL replied:
it's not. Just questioning the introduction of an experimental interface.

Maybe we can extract better API over time without being stuck on something.

ro0NL replied:
from https://symfony.com/doc/current/contributing/code/experimental.html

> gather feedback from real-world usage, adapt the API, or remove it altogether

i'd say

_gather feedback from real-world usage, adapt the API, or add one_ :)

nicolas-grekas replied:
Oh OK!
I think that while it's experimental, we should still try to achieve something that is as close as possible to what we would consider stable.
Experimental is here to give a simple path when we realize we made some mistakes along this path.
With this approach, adding the interface makes sense to me.
The corollary here is of course that we should review the PR (and this interface) as if it were not experimental.

ro0NL replied:
my point is, we can always extract and add an interface, but not always remove one.

So why pushing the interface now, when technically we can do without still (IMHO, or change my mind ;))

nicolas-grekas replied:
We can remove it, because experimental. Yet I don't think we will, because the review will confirm we need it I believe. That's why we should add it.

ro0NL replied:
let me rephrase it, after the experimental flag is gone, removing is more hassle than adding.

For now im not convinced this interface is needed, but it's not a blocker either.

tgalopin replied:
My personal POV is relative usage: if we don't add the interface early on, people will typehint the class implementation, meaning we will need to provide a autowirable service until 7.0. I don't think that's wise: we should always promote the usage of interfaces in typehints IMO, esp when they are linked to BC.

And if we consider the autowirable service experimental as well (allowing us to remove it earlier than 7.0), then we don't have any reason not to use an interface I think.

ro0NL replied:
this is unrelated to autowiring. HtmlSanitizer is as autowire-able as HtmlSanitizerInterface.

typehinting interfaces over classes is not a rule by law ;)

the only reason for an interface, technically spoken, is if we expect some form of polymorphism. But generally i believe we do not (eg. in tests you can actually just sanitize HTML isnt it).

nicolas-grekas replied:
Software design should allow for what is *not* expected (aka we should not have to anticipate future needs for the present design to be correct.) We have SOLID engineering rules on the topic that remind us that this is achieved by type-hinting against abstractions (=interfaces in PHP, for services at least).
That's why the interface makes sense to me. I'm for keeping it, and autowiring it also.

ro0NL replied:
sound reasonable :+1: (still not convinced though :))

Let's settle.

stof replied:
Well, if we want to introduce a TraceableSanitizer in the future to hook into the debug tools, the interface is a must-have to allow us to implement that through decoration rather than inheritance. If we initially push projects to typehint the class, we cannot use decoration on the service.

ro0NL replied:
i'd be curious how much sense a system-wide HtmlSanitizerConfig makes anyway :/ (or SF framework managing the instances for that matter)

tigitz commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizer.php:
```diff
@@ -0,0 +1,131 @@
+        return '' === $html || preg_match('//u', $html);
+    }
+
+    private function createDomVisitorForContext(string $context)
```
```suggestion
    private function createDomVisitorForContext(string $context): DomVisitor
```

tigitz commented on src/Symfony/Component/HtmlSanitizer/Visitor/DomVisitor.php:
```diff
@@ -0,0 +1,153 @@
+        return $cursor->node;
+    }
+
+    private function visitNode(\DOMNode $domNode, Cursor $cursor)
```
```suggestion
    private function visitNode(\DOMNode $domNode, Cursor $cursor): void
```

tigitz commented on src/Symfony/Component/HtmlSanitizer/Visitor/DomVisitor.php:
```diff
@@ -0,0 +1,153 @@
+        $cursor->node = $cursor->node->getParent();
+    }
+
+    private function enterNode(string $domNodeName, \DOMNode $domNode, Cursor $cursor)
```
```suggestion
    private function enterNode(string $domNodeName, \DOMNode $domNode, Cursor $cursor): void
```

tigitz commented on src/Symfony/Component/HtmlSanitizer/Visitor/DomVisitor.php:
```diff
@@ -0,0 +1,153 @@
+        $cursor->node = $node;
+    }
+
+    private function visitChildren(\DOMNode $domNode, Cursor $cursor)
```
```suggestion
    private function visitChildren(\DOMNode $domNode, Cursor $cursor): void
```

tigitz commented on src/Symfony/Component/HtmlSanitizer/Visitor/DomVisitor.php:
```diff
@@ -0,0 +1,153 @@
+    /**
+     * Set attributes from a DOM node to a sanitized node.
+     */
+    private function setAttributes(string $domNodeName, \DOMNode $domNode, Node $node, array $allowedAttributes = [])
```
```suggestion
    private function setAttributes(string $domNodeName, \DOMNode $domNode, Node $node, array $allowedAttributes = []): void
```

tigitz reviewed (changes requested):
Some missing return types

tigitz commented:
General question, is it planned to support only specific browser versions ?

I'm asking this because `<BGSOUND SRC="javascript:alert('XSS');">` (found [here](https://cheatsheetseries.owasp.org/cheatsheets/XSS_Filter_Evasion_Cheat_Sheet.html#bgsound)) for example is not tested (so not supported I guess ?) but seems to be problematic only for IE AFAIU (https://developer.mozilla.org/fr/docs/Web/HTML/Element/bgsound).

So it could be legitimately ignored because it's an outdated browser but library consumers that still needs to support those browsers (and they exist) will probably require the package and won't realise it actually does not cover their needs **by design**.

If that's the plan I would recommend to be more upfront about the supported browsers and versions, whether it's phpdoc or doc itself.

tgalopin commented:
> General question, is it planned to support only specific browser versions ?

I think the aim is to be as safe as possible, ie. ensuring we remove all sources of XSS for all browsers.

You raise an important question though: I'm the one who originally explored sources of potential XSS evasions, and I may have missed some. Your example is a good one: `<BGSOUND SRC="javascript:alert('XSS');">` was indeed missing (because it wasn't listed on https://www.w3schools.com/tags/att_src.asp and I used this reference for the attribute sanitizer).

I'm going to explore https://cheatsheetseries.owasp.org/cheatsheets/XSS_Filter_Evasion_Cheat_Sheet.html in more details (thanks for the link!) but I think if people reviewing this PR could also have a look at potentially missed XSS evasion strategies, that's be great!

tgalopin commented:
@tigitz I added a test and a sanitization for your case FYI.

@nicolas-grekas I just added the ability to call `->forceAttribute('a', 'rel', 'noopener noreferrer')` to force the value of an attribute on an element.

@stof 
> should we allow using * inside the array too, in case we want to allow all standard attributes plus some custom ones ?

I guess one could easily call `allowAttribute` right after? I think it would complexify a bit the code without much added benefit but I can do it if it's a deal breaker.

ro0NL commented on src/Symfony/Component/HtmlSanitizer/Parser/ParserInterface.php:
```diff
@@ -0,0 +1,28 @@
+{
+    /**
+     * Parse a given string and returns a DOMNode tree.
+     * This method must return null if the string cannot be parsed as HTML.
```
How does it account for error handling?

tgalopin replied:
See https://github.com/symfony/symfony/pull/44681#discussion_r771536972

ro0NL replied:
the masterminds parser collects errors, though. Not sure hiding them is wise.

perhaps from a sanitazation POV it is :+1:

tgalopin replied:
It's now possible to inject a custom parser to gather errors if one wishes.

tigitz commented:
> I think the aim is to be as safe as possible, ie. ensuring we remove all sources of XSS for all browsers.

Understood. I guess you evaluate the performance penalty of doing unnecessary "outdated browsers" sanitation checks in a "modern browsers" context to be marginal ?

> if people reviewing this PR could also have a look at potentially missed XSS evasion strategies, that's be great!

Sure! For completeness, I found the given link through test suites of other open-source html sanitization libs. Could be useful to increase case coverage:

- https://github.com/mganss/HtmlSanitizer/blob/master/test/HtmlSanitizer.Tests/Tests.cs
- https://github.com/apostrophecms/sanitize-html/blob/main/test/test.js
- https://github.com/rgrove/sanitize/tree/main/test

nicolas-grekas commented on src/Symfony/Component/HtmlSanitizer/Parser/MastermindsParser.php:
```diff
@@ -0,0 +1,27 @@
+{
+    public function parse(string $html): ?\DOMNode
+    {
+        return (new HTML5())->loadHTMLFragment($html);
```
missing constructor to allow defining the $defaultOptions argument of the HTML5 class?

nicolas-grekas commented on src/Symfony/Component/HtmlSanitizer/Visitor/Node/Node.php:
```diff
@@ -0,0 +1,106 @@
+        return $this->attributes[$name] ?? null;
+    }
+
+    public function setAttribute(string $name, ?string $value)
```
`: void`

nicolas-grekas commented on src/Symfony/Component/HtmlSanitizer/Visitor/Node/NodeInterface.php:
```diff
@@ -0,0 +1,39 @@
+namespace Symfony\Component\HtmlSanitizer\Visitor\Node;
+
+/**
+ * A node represents the sanitized version of a DOM node in the sanitized tree.
```
Represents the sanitized version of a DOM node in the sanitized tree.

the rest after a blank line

all phpdocs should have a title + an optional description. right now, many use a long title instead

nicolas-grekas commented on src/Symfony/Component/HtmlSanitizer/composer.json:
```diff
@@ -0,0 +1,31 @@
+{
+    "name": "symfony/html-sanitizer",
+    "type": "library",
+    "description": "Provides an object-oriented API to sanitize untrusted HTML input for safe insertion into a document’s DOM.",
```
`s/’/'`

nicolas-grekas commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerInterface.php:
```diff
@@ -0,0 +1,47 @@
+
+/**
+ * Provides an object-oriented API to sanitize untrusted HTML input for safe
+ * insertion into a document’s DOM.
```
`s/’/'/`

nicolas-grekas commented on src/Symfony/Component/HtmlSanitizer/README.md:
```diff
@@ -0,0 +1,106 @@
+=======================
+
+The HtmlSanitizer component provides an object-oriented API to sanitize
+untrusted HTML input for safe insertion into a document’s DOM.
```
please replace all unicode quotes by an ascii single quote

tgalopin commented:
I pushed more cases coming from https://cheatsheetseries.owasp.org/cheatsheets/XSS_Filter_Evasion_Cheat_Sheet.html#bgsound (thanks @tigitz !) and I refactored the default attribute sanitizer to check all elements for URL attributes (`'src', 'href', 'lowsrc', 'background', 'ping'`). It should be more reliable and safe than having a manual list.

ro0NL commented on src/Symfony/Component/HtmlSanitizer/Parser/MastermindsParser.php:
```diff
@@ -0,0 +1,35 @@
+
+    public function parse(string $html): ?\DOMNode
+    {
+        if (!$this->parser) {
+            $this->parser = new HTML5();
+        }
```
`??=`

derrabus commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizer.php:
```diff
@@ -0,0 +1,131 @@
+    private ParserInterface $parser;
+
+    /**
+     * @var DomVisitor[]
```
```suggestion
     * @var array<string, DomVisitor>
```

derrabus commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,469 @@
+    /**
+     * Links schemes that should be retained, other being dropped.
+     *
+     * @var string[]
```
```suggestion
     * @var list<string>
```

derrabus commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,469 @@
+     * Allows only a given list of schemes to be used in links href attributes.
+     *
+     * All other schemes will be dropped.
+     */
```
```suggestion
     *
     * @param list<string> $allowLinkSchemes
     */
```

derrabus commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,469 @@
+    }
+
+    /**
+     * @return string[]
```
```suggestion
     * @return list<string>
```

stof replied:
if we don't care whether the array is actually a list (because we never use the keys), should we enforce it to be a list ? psalm is strict about it.

derrabus commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,469 @@
+    /**
+     * Links hosts that should be retained (by default, all hosts are allowed).
+     *
+     * @var string[]|null
```
```suggestion
     * @var list<string>|null
```

derrabus commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,469 @@
+     *
+     * All other hosts will be dropped. By default all hosts are allowed
+     * ($allowedLinkHosts = null).
+     */
```
```suggestion
     *
     * @param list<string>|null $allowLinkHosts
     */
```

derrabus commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,469 @@
+    /**
+     * Image/Audio/Video schemes that should be retained, other being dropped.
+     *
+     * @var string[]
```
```suggestion
     * @var list<string>
```

derrabus commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,469 @@
+     * Allows only a given list of schemes to be used in media source attributes (img, audio, video, ...).
+     *
+     * All other schemes will be dropped.
+     */
```
```suggestion
     *
     * @param list<string> $allowedMediaSchemes
     */
```

derrabus commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,469 @@
+    }
+
+    /**
+     * @return string[]
```
```suggestion
     * @return list<string>
```

derrabus commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,469 @@
+    }
+
+    /**
+     * @return string[]|null
```
```suggestion
     * @return list<string>|null
```

derrabus commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,469 @@
+    /**
+     * Image/Audio/Video hosts that should be retained (by default, all hosts are allowed).
+     *
+     * @var string[]|null
```
```suggestion
     * @var list<string>|null
```

derrabus commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,469 @@
+     *
+     * All other hosts will be dropped. By default all hosts are allowed
+     * ($allowMediaHosts = null).
+     */
```
```suggestion
     *
     * @param list<string>|null $allowedMediaHosts
     */
```

derrabus commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,469 @@
+    }
+
+    /**
+     * @return string[]|null
```
```suggestion
     * @return list<string>|null
```

derrabus commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,469 @@
+    /**
+     * Sanitizers that should be applied to specific attributes in addition to standard sanitization.
+     *
+     * @var AttributeSanitizerInterface[]
```
```suggestion
     * @var list<AttributeSanitizerInterface>
```

derrabus commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,469 @@
+    }
+
+    /**
+     * @return AttributeSanitizerInterface[]
```
```suggestion
     * @return list<AttributeSanitizerInterface>
```

derrabus commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -0,0 +1,469 @@
+    public function withoutAttributeSanitizer(AttributeSanitizerInterface $sanitizer): static
+    {
+        $clone = clone $this;
+        unset($clone->attributeSanitizers[array_search($sanitizer, $clone->attributeSanitizers, true)]);
```
```suggestion
        $clone->attributeSanitizers = array_values(array_filter(
            $this->attributeSanitizers,
            static fn ($current) => $current !== $sanitizer
        ));
```

Nit-picking: This way we make sure, the array is a list again after removing the item.

derrabus reviewed (commented):
Let's document list arrays a bit better.

tgalopin commented:
I fixed all reviews :)

nicolas-grekas commented on src/Symfony/Component/HtmlSanitizer/Parser/MastermindsParser.php:
```diff
@@ -0,0 +1,33 @@
+ */
+final class MastermindsParser implements ParserInterface
+{
+    public function __construct(private ?HTML5 $parser = null)
```
I'm afraid this won't work because HTML5 is stateful...

tgalopin commented:
Updated

norkunas commented on src/Symfony/Component/HtmlSanitizer/TextSanitizer/UrlSanitizer.php:
```diff
@@ -0,0 +1,136 @@
+
+    private static function isHostlessScheme(?string $scheme): bool
+    {
+        return \in_array($scheme, ['data', 'mailto', 'tel', 'file', 'view-source', 'maps', 'chrome', 'blob'], true);
```
https://en.wikipedia.org/wiki/Geo_URI_scheme `geo` is also hostless :)

nicolas-grekas commented on src/Symfony/Component/HtmlSanitizer/TextSanitizer/StringSanitizer.php:
```diff
@@ -0,0 +1,65 @@
+     */
+    public static function encodeHtmlEntities(string $string): string
+    {
+        return str_replace(
```
processing consts is wasted CPU
let's use strtr or turn REPLACEMENTS into an array that is directly usable by str_replace

nicolas-grekas reviewed (approved):
I'm fine with the component. I guess you want to proceed with FWB integration in another PR?

tgalopin commented:
@nicolas-grekas I'm working on FWB integration but we could indeed open a new PR for that. I don't mind either way

tgalopin commented:
Let's open a different PR, it'll ease reviewing of the configuration/FW integration without having too much comments here.

@derrabus are you OK with the component now?

tgalopin commented:
Just opened https://github.com/symfony/symfony/pull/44798 for the FWB integration

fabpot reviewed (approved):
Let's merge early to gather feedback.
Well done @tgalopin

fabpot commented:
Thank you @tgalopin.

DocFX commented:
So happy to see this one integrated in SF! Thanks for the job, you all rock. <3
